### ISSUE TITLE
A problem in MC Prediction Solution

### ISSUE THREAD

**Issue by dslwz2008** [dennybritz/reinforcement-learning#149]
Recently I read your codes in  MC Prediction Solution.ipynb and found that there may be a problem in the mc_prediction function.  You calculate average returns for each state like this:

`            # Calculate average return for this state over all sampled episodes
            returns_sum[state] += G
            returns_count[state] += 1.0
            V[state] = returns_sum[state] / returns_count[state]
`
I have a question here: whether or not the same state may exist in different episodes？ I think it is possible. If so, the average return V can not be calculated in this way and the algorithm of average calculating may be a little different. Or, you should calculate the V outside the for loop before return it.

**Comment by jonahweissman** [comment]
I'm not sure I understand your question, but that code will definitely find the average return for each state. The goal is to average returns from the same state over many episodes. States may appear multiple times in an episode, but the code only averages the first-visit returns.

**Comment by bonifacechacha** [comment]
I found the same issue on the line
 `V[state] = returns_sum[state] / returns_count[state]`
This line is suppose to be out of the loop if we are calculating the average over a number of visit from the episode

**Comment by levindabhi** [comment]
Yes . I also think it should be outside but it doesn't make any difference as V[state] value after the loop is same as last iteration of loop . May be @dennybritz  put that line inside the loop to use state variable which is in loop .

**Comment by bonifacechacha** [comment]
I can see that too. The logic is still correct as it is

**Comment by disadone** [comment]
Some states may not occur in every episodes, so you have to write the statement inside.
